Sources/GRPC/GRPCServerCodec.swift

```diff
@@ -75,8 +74,9 @@ extension GRPCServerCodec: ChannelOutboundHandler {
 
     case .message(let message):
       do {
-        let messageData = try message.serializedData()
-        context.write(self.wrapOutboundOut(.message(messageData)), promise: promise)
+        var buffer = ByteBufferAllocator().buffer(capacity: 0)
+        try message.serialize(into: &buffer)
+        context.write(self.wrapOutboundOut(.message(buffer.readData(length: buffer.readableBytes)!)), promise: promise)
```

Comment:
I would gladly do it but do you have a structure in mind or should I look into how we can do it?

Reply:
Sure, I'll try to explain the problem and what needs to be done to improve it.

### Problem

As you know, before this change we explicitly expected payloads to be `Message` from SwiftProtobuf. The only serialization and deserialization methods from `Message` are via `Data` from Foundation. Internally we use NIO and our currency type for storing bytes is NIOs `ByteBuffer`.

Because we always expect a `Data` from `Message`, the `write` method on `LengthPrefixedMessageWriter` accepts a `Data` and writes into a `ByteBuffer`.

With the protocol we always serialize into a `ByteBuffer` (for `Message` we _have_ to go via `Data` first). This means that we've gone from:

1. `Message` to `Data` (serialization)
2. `Data` to `ByteBuffer` (length prefixing)

to

1. `Message` to `Data` (serialization) 
2. `Data` to `ByteBuffer` (we need to for `GRPCPayload`)
3. `ByteBuffer` to `Data` (we need to for `LengthPrefixedMessageWriter.write`)
4. `Data` to `ByteBuffer` (length prefixing)

Clearly we don't want to copy bytes from `Data` to `ByteBuffer` and back from `ByteBuffer` to `Data` again (and we also don't want our API for `GRPCPayload` to be defined by `Data`).

Now I need to provide some background on the `LenthPrefixedMessageWriter`: gRPC payloads are sent with a 5-byte prefix (I might have explained this before but I'll repeat it for completeness). The first byte indicates whether the payload is compressed or not, the next four bytes indicate the length of the payload (the compressed length if it is compressed). 

If the payload isn't compressed then this is actually really straightforward: we write a `0`-byte we write the length of the payload as a `UInt32` and then we write the payload. For compression we have to write the a `1`-byte, leave a 4-byte gap for the length, compress the payload into the buffer and not how many bytes it was and then fill in the 4-byte gap.

### Possible Solution

So I think the best we can do now (I'll use protobuf as an example again) is actually to pre-allocate the extra 5-byte space at the beginning of the buffer before we pass it into `message.serialize(into:)`.

If we aren't using compression then all we need to do is fill in the first 5 bytes, easy! (And we have the `Message` to `Data` to `ByteBuffer`, as we had before, for Flatbuffers I expect we have something similar.)

When we use compression things are slightly different because we can't compress in place, so we have an extra buffer but in this case the same applies, we can avoid the extra hop via `Data`.

Hopefully this all makes sense so far?

### What needs to be done?

- `LengthPrefixedMessageWriter.write` accepts a `Data`, instead I think the writer should accept a `GRPCPayload` (as the writer knows whether compression is being used so knows how to fill in the first 5-bytes of the payload etc, it can also choose whether it _should_ preallocate 5 bytes, if we're using compression then we don't need to since we have to compress into a different buffer anyway).
- `Zlib.deflate` also accepts a `Data` if I recall correctly, this will need to be changed to a `ByteBuffer`

On the server side of things, this will require having to move some components around. At the moment the length prefixed writer lives in `HTTP1ToRawGRPCServerCodec`; this will have to move to `GPRCServerCodec`. This means the types that get sent between these handlers will also have to change (`_RawGRPCServerResponsePart.message` will have a `ByteBuffer` as its payload which will be a length prefixed message). I also expect a number of tests will have to be updated to support these changes.